### PR TITLE
Adding multi lang tts

### PR DESCRIPTION
**Pull request by TonnyAn**

### CODE REVIEW

File: app/src/main/java/bellevuecollege/edu/cookpal/recipe_details/RecipeDetailsFragment.kt [hunk 1]
```diff
@@ -8,32 +9,58 @@ import android.util.Log
...
-                mTTS.language = Locale.UK
-            }
-        }
-    }
+    /**
+     * Can remove onCreate function
+     * spinner is default to english so
+     * TTS engine is created there.
+     * See OnItemSelected function down below.
+     */
```
Comment (DavePie) [inline app/src/main/java/bellevuecollege/edu/cookpal/recipe_details/RecipeDetailsFragment.kt:50]: Is Trang using this? If not, then can this be removed?
I'll ask Trang to review this and comment if you can remove these portions or not.

**Review by elliotwe — APPROVED** [review]

File: app/src/main/java/bellevuecollege/edu/cookpal/recipe_details/RecipeDetailsFragment.kt [hunk 2]
```diff
@@ -50,17 +77,46 @@ class RecipeDetailsFragment : Fragment() {
...
         val tempView = binding.viewModel
+        //Get selected recipe information and place into xml
         if (tempView != null) {
-
             tempView.selectedRecipe.observe(viewLifecycleOwner) { parsedRecipe ->
-                binding.recipeSummary.text = parsedRecipe.summary
-                binding.recipeIngredients.text = parsedRecipe.ingredients.toString()
-                binding.recipeInstructions.text = parsedRecipe.steps.toString()
+                recipe = parsedRecipe
+                //@TODO figure out what to do about recipe summaries
```
Comment (DavePie) [inline app/src/main/java/bellevuecollege/edu/cookpal/recipe_details/RecipeDetailsFragment.kt:90]: Make sure to bring this up next meeting

File: app/src/main/java/bellevuecollege/edu/cookpal/recipe_details/RecipeDetailsFragment.kt [hunk 3]
```diff
@@ -8,32 +9,58 @@ import android.util.Log
...
+//            activity?.applicationContext
+//        ) { status ->
+//            if (status != TextToSpeech.ERROR) {
+//                //if there is no error then set language
+//                mTTS.language = Locale.UK
+//            }
+//        }
+    //}
 
     override fun onCreateView(
```
Comment (DavePie) [inline app/src/main/java/bellevuecollege/edu/cookpal/recipe_details/RecipeDetailsFragment.kt:65]: This is quite a long method making it hard to parse. Are there any subdivisions you can do?

File: app/src/main/java/bellevuecollege/edu/cookpal/recipe_details/RecipeDetailsFragment.kt [hunk 4]
```diff
@@ -50,17 +77,46 @@ class RecipeDetailsFragment : Fragment() {
...
-                binding.recipeIngredients.text = parsedRecipe.ingredients.toString()
-                binding.recipeInstructions.text = parsedRecipe.steps.toString()
+                recipe = parsedRecipe
+                //@TODO figure out what to do about recipe summaries
+                //binding.recipeSummary.text = parsedRecipe.summary
+                binding.recipeIngredients.text =
+                    parsedRecipe.ingredients.joinToString("") { "- $it\n" }
+                binding.recipeInstructions.text = parsedRecipe.steps.mapIndexed{index, s -> "${index+1}) $s" }.joinToString("") { "$it\n" }
+
+                //@TODO Old instructions code, may need to remove
```
Comment (DavePie) [inline app/src/main/java/bellevuecollege/edu/cookpal/recipe_details/RecipeDetailsFragment.kt:96]: Trang: Keep/Remove?

File: app/src/main/java/bellevuecollege/edu/cookpal/recipe_details/RecipeDetailsFragment.kt [hunk 5]
```diff
@@ -159,6 +215,196 @@ class RecipeDetailsFragment : Fragment() {
         return binding.root
     }
 
+    /**
+     * Spinner drop down listeners
+     */
+    override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
+        when (position) {
+            0 -> {
```
Comment (DavePie) [inline app/src/main/java/bellevuecollege/edu/cookpal/recipe_details/RecipeDetailsFragment.kt:223]: Would be nice to have enums here instead of numbers to be a bit more descriptive. Strings work too.
General example:
https://riptutorial.com/kotlin/example/20008/when-statement-with-enums

File: app/src/main/java/bellevuecollege/edu/cookpal/recipe_details/RecipeDetailsFragment.kt [hunk 6]
```diff
@@ -159,6 +215,196 @@ class RecipeDetailsFragment : Fragment() {
...
+                        //if there is no error then set language
+                        mTTS.language = espanol
+                    }
+                }
+
+                /**
+                 * Set up spanish translator model
+                 */
+                val options = TranslatorOptions.Builder()
+                    .setSourceLanguage(TranslateLanguage.ENGLISH)
```
Comment (DavePie) [inline app/src/main/java/bellevuecollege/edu/cookpal/recipe_details/RecipeDetailsFragment.kt:253]: Why do you set it to English first?

File: app/src/main/java/bellevuecollege/edu/cookpal/recipe_details/RecipeDetailsFragment.kt [hunk 7]
```diff
@@ -159,6 +215,196 @@ class RecipeDetailsFragment : Fragment() {
...
+                            }
+                    }
+                    .addOnFailureListener {
+                        //Model couldn't be downloaded or other internal error
+                        Log.e("Translator", "Model download failed")
+                    }
+            }
+            2 -> {
+                //Set up TTS language
+                mTTS = TextToSpeech(
```
Comment (DavePie) [inline app/src/main/java/bellevuecollege/edu/cookpal/recipe_details/RecipeDetailsFragment.kt:280]: Lots of repetition. Is it possible to put it in a method with language as parameter?

**Review by DavePie — CHANGES_REQUESTED** [review]

**Review by DavePie — APPROVED** [review]

File: app/src/main/java/bellevuecollege/edu/cookpal/recipe_details/RecipeDetailsFragment.kt [hunk 8]
```diff
@@ -50,17 +77,46 @@ class RecipeDetailsFragment : Fragment() {
...
+            up.setProfile(data)
+        }
+
+        //View model
         val tempView = binding.viewModel
+        //Get selected recipe information and place into xml
         if (tempView != null) {
-
             tempView.selectedRecipe.observe(viewLifecycleOwner) { parsedRecipe ->
-                binding.recipeSummary.text = parsedRecipe.summary
```
Comment (trang2020) [inline app/src/main/java/bellevuecollege/edu/cookpal/recipe_details/RecipeDetailsFragment.kt:57]: If you remove recipeSummary, you need to update the Summary in view (recipe_details_fragment) accordingly

File: .idea/misc.xml [hunk 9]
```diff
@@ -60,7 +60,7 @@
         <entry key="..\:/Users/Tonny An/AndroidStudioProjects/CookPal/app/src/main/res/layout/grid_row_recipe.xml" value="0.362962962962963" />
         <entry key="..\:/Users/Tonny An/AndroidStudioProjects/CookPal/app/src/main/res/layout/home_screen_fragment.xml" value="0.33" />
         <entry key="..\:/Users/Tonny An/AndroidStudioProjects/CookPal/app/src/main/res/layout/horizontal_grid_recipe.xml" value="0.19220779220779222" />
-        <entry key="..\:/Users/Tonny An/AndroidStudioProjects/CookPal/app/src/main/res/layout/recipe_details_fragment.xml" value="0.6730290456431536" />
```
Comment (trang2020) [inline .idea/misc.xml:63]: Do not add .idea folder to repository, it is project and machine dependent.

File: app/build.gradle [hunk 10]
```diff
@@ -24,11 +24,11 @@ apply plugin: 'com.google.gms.google-services'
 
 
 android {
-    compileSdkVersion 28
+    compileSdkVersion 31
```
Comment (trang2020) [inline app/build.gradle:27]: There is an issue with TTS support in SDK > 28. If we use SDK 31, we have to run/test code on Emulator with API <= 28, otherwise TTS will not work.

File: app/src/main/java/bellevuecollege/edu/cookpal/recipe_details/RecipeDetailsFragment.kt (hunk 1 above)
Comment (trang2020) [reply app/src/main/java/bellevuecollege/edu/cookpal/recipe_details/RecipeDetailsFragment.kt:49]: Remove If TTS works as expected.

File: app/src/main/java/bellevuecollege/edu/cookpal/recipe_details/RecipeDetailsFragment.kt (hunk 4 above)
Comment (trang2020) [reply app/src/main/java/bellevuecollege/edu/cookpal/recipe_details/RecipeDetailsFragment.kt:104]: Loosing Summary is a bummer. We should add an If clause to display Summary If it exists.

File: app/src/main/java/bellevuecollege/edu/cookpal/recipe_details/RecipeDetailsFragment.kt [hunk 11]
```diff
@@ -159,6 +223,103 @@ class RecipeDetailsFragment : Fragment() {
...
+                            .build()
+                    }
+                    //Create language model
+                    textTranslator = options?.let { Translation.getClient(it) }!!
+                    var conditions = DownloadConditions.Builder()
+                        .requireWifi()
+                        .build()
+                    //download language model
+                    //Language models are around 30MB, don't have too many.
+                    textTranslator.downloadModelIfNeeded(conditions)
```
Comment (trang2020) [inline app/src/main/java/bellevuecollege/edu/cookpal/recipe_details/RecipeDetailsFragment.kt:303]: Do we download models every time we need to translate anything? Can we cache models?

File: app/src/main/java/bellevuecollege/edu/cookpal/recipe_details/RecipeDetailsFragment.kt [hunk 12]
```diff
@@ -159,6 +223,103 @@ class RecipeDetailsFragment : Fragment() {
...
+                        .addOnSuccessListener {
+                            // Model downloaded successfully. Okay to start translating
+                            // Set a flag, unhide the translation UI, etc.
+                            Log.d("Translator", "Model download successful")
+                            textTranslator.translate(fullRecipe)
```
Comment (trang2020) [inline app/src/main/java/bellevuecollege/edu/cookpal/recipe_details/RecipeDetailsFragment.kt:308]: Does it block the UI? Is it CPU heavy? Should we cache it for future queries?
